src/main/jbake/content/blog/2017/06/23/opentracing-appmetrics.adoc

```diff
@@ -0,0 +1,270 @@
+invoking an external service. They provide an ideal basis upon which to report metrics
+(count and duration) regarding these points within a service.
+
+Therefore a new https://github.com/opentracing-contrib/java-metrics[OpenTracing contrib project] has
+been established (initially just for Java) to intercept
+the finished spans, and create the relevant metrics. These metrics are then submitted to a _MetricsReporter_
+for recording - the initial implementation of this interface being for https://prometheus.io/[Prometheus].
+
+The first step is to expose an endpoint for collecting the Prometheus metrics. Each service has the following
+configuration:
+
+----
+@Configuration
+@ConditionalOnClass(CollectorRegistry.class)
+public class PrometheusConfiguration {
+
+     @Bean
+     @ConditionalOnMissingBean
+     CollectorRegistry metricRegistry() {
+         return CollectorRegistry.defaultRegistry;
+     }
+
+     @Bean
+     ServletRegistrationBean registerPrometheusExporterServlet(CollectorRegistry metricRegistry) {
+           return new ServletRegistrationBean(new MetricsServlet(metricRegistry), "/metrics");
+     }
+}
+----
+
+This will allow the Prometheus metrics to be obtained from the service's `/metrics` REST endpoint.
+
+Each service then requires a configuration to obtain the `io.opentracing.Tracer`: 
+
+----
+@Configuration
+public class TracerConfiguration implements javax.servlet.ServletContextListener {
+
+	@Bean
+	public io.opentracing.Tracer tracer() {
+		return io.opentracing.contrib.metrics.Metrics.decorate(
+			io.opentracing.contrib.tracerresolver.TracerResolver.resolveTracer(),
+			PrometheusMetricsReporter.newMetricsReporter()
+				.withBaggageLabel("transaction","n/a")
+				.build());
+	}
+
+	@Override
+	public void contextInitialized(javax.servlet.ServletContextEvent sce) {
+		sce.getServletContext().setAttribute(io.opentracing.contrib.web.servlet.filter.TracingFilter.SKIP_PATTERN, Pattern.compile("/metrics"));
+	}
+
+	...
+----
+
+The first method uses the
+https://github.com/opentracing-contrib/java-tracerresolver[TracerResolver] to provide a vendor
+neutral approach for accessing a `Tracer`. This tracer is then decorated with the metrics capability
+using a `PrometheusMetricsReporter`. This metrics reporter is further configured to add a special
+label related to the baggage key `transaction` (discussed later).
+
+By default, the Servlet OpenTracing integration will trace all REST endpoints. Therefore in the
+second method above we add an attribute that will inform the instrumentation to ignore the `/metrics`
+endpoint. Otherwise will we have tracing data reported each time Prometheus reads the metrics for the
+service.
+
+
+== Deploying on Kubernetes
+
+The steps to setup an environment on Kubernetes is discussed in the
+https://github.com/objectiser/opentracing-prometheus-example[example codebase]. A summary of the steps is:
+
```

Comment:
One word (setup) as an adjective or noun, two words (set up) as a verb. (I almost always have to look it up.)  Since it’s a verb here, “set up” should be two words.